src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
-    if (build_coarse_relation)
+    return sec_M.Collect();
+}
+
+void GraphCoarsen::ComputeEdgeTargets(const MixedMatrix& mgl,
+                                      const ParMatrix& face_perm_edge)
+{
+    const SparseMatrix& face_edge = face_perm_edge.GetDiag();
+
+    auto shared_sigma = CollectSigma(face_edge);
+    auto shared_M = CollectM(mgl.LocalM());
+    auto shared_D = CollectD(mgl.LocalD());
+
+    const SparseMatrix& face_shared = gt_.face_face_.GetOffd();
+
+    SharedEntityComm<DenseMatrix> sec_face(gt_.face_true_face_);
+    DenseMatrix collected_sigma;
+
+    int num_faces = gt_.face_edge_local_.Rows();
+
+    for (int face = 0; face < num_faces; ++face)
     {
-        Agg_dof_i = new int[nAggs + 1];
-        Agg_dof_i[0] = 0;
+        int num_face_edges = face_edge.RowSize(face);
+
+        if (!sec_face.IsOwnedByMe(face))
+        {
+            edge_targets_[face].SetSize(num_face_edges, 0);
+            continue;
+        }
+
+        if (num_face_edges == 1)
+        {
+            edge_targets_[face] = DenseMatrix(1, 1, {1.0});
+            continue;
+        }
+
+        const auto& face_M = shared_M[face];
+        const auto& face_D = shared_D[face];
+        const auto& face_sigma = shared_sigma[face];
+
+        linalgcpp::HStack(face_sigma, collected_sigma);
+
+        bool shared = face_shared.RowSize(face) > 0;
+
+        // TODO(gelever1): resolve this copy.  (Types must match so face_? gets copied and promoted
+        // to rvalue).
+        const auto& M_local = shared ? Combine(face_M, num_face_edges) : face_M[0];
+        const auto& D_local = shared ? Combine(face_D, num_face_edges) : face_D[0];
+        const int split = shared ? face_D[0].Rows() : GetSplit(face);
+
+        GraphEdgeSolver solver(M_local, D_local);
+        Vector one_neg_one = MakeOneNegOne(D_local.Rows(), split);
+
+        Vector pv_sol = solver.Mult(one_neg_one);
+        VectorView pv_sigma(pv_sol.begin(), num_face_edges);
+
+        edge_targets_[face] = Orthogonalize(collected_sigma, pv_sigma, 0, max_evects_);
     }
 
-    // compute nnz in each row (fine edge)
-    int* Pedges_i = new int[nedges + 1]();
-    for (unsigned int i = 0; i < nAggs; i++)
+    sec_face.Broadcast(edge_targets_);
+
+    ScaleEdgeTargets(mgl.LocalD());
+}
+
+void GraphCoarsen::ScaleEdgeTargets(const SparseMatrix& D_local)
+{
+    int num_faces = gt_.face_edge_.Rows();
+
+    for (int face = 0; face < num_faces; ++face)
     {
-        GetTableRow(Agg_edge, i, local_fine_dofs);
-        GetTableRow(Agg_face, i, faces);
-        nlocal_coarse_dofs = vertex_target[i].Width() - 1;
-        for (int j = 0; j < faces.Size(); ++j)
-            nlocal_coarse_dofs += edge_traces[faces[j]].Width();
-        for (int j = 0; j < local_fine_dofs.Size(); ++j)
-            Pedges_i[local_fine_dofs[j] + 1] = nlocal_coarse_dofs;
-        if (build_coarse_relation)
+        DenseMatrix& edge_traces(edge_targets_[face]);
+        if (edge_traces.Cols() < 1)
```

Comment:
We need to always have `edge_traces.Cols() >= 1`?

Reply:
Yes, this should probably assert the condition instead of continuing.